ReactWindows/ReactNative.Shared/UIManager/ReactShadowNode.cs

```diff
@@ -1191,18 +1211,33 @@ public int GetNativeOffsetForChild(ReactShadowNode child)
             return index;
         }
 
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
         /// <summary>
         /// Disposes the shadow node.
         /// </summary>
-        public void Dispose()
+        protected virtual void Dispose(bool disposing)
         {
+            if (_disposed)
+            {
+                return;
+            }
+
             YogaNode nodeToDispose = (IsDelegatedLayout && _initialYogaNode != null) ? _initialYogaNode : _yogaNode;
 
             if (nodeToDispose != null)
             {
+                RemoveAndDisposeAllChildren();
+                UnlinkFromParent();
                 nodeToDispose.Reset();
```

Comment:
I think you should skip touching managed references if you got called with "disposing == false" (from finalizer), the objects are finalized out of some order.

Reply:
To clarify, based on my rusty memories from times where the only .NET programming was the interop with Win32 programming, the idea is that:
- if Dispose is executed at normal times (so disposing == true), you may clean the native references and also release various managed references, etc. since that proactively releases sets of objects to GC without depending on some root be released.
- if Dispose is executed at finalizing time (so disposing == false), you can't touch any references anymore since objects are finalized on some random order (there's no "correct & safe" order when you have to finalize reference cycles). So in this mode you only release native references.

In fact that's the reason the "disposing" parameter was added to the pattern.

In the current matter:
- RemoveAnddisposeAllChildren touches managed references, shouldn't be executed in that "finalizing" case
- same for UnlinkFromParent.
now when it gets to those wrappers, object pools, etc., it becomes harder to grasp, if you go by the book nothing can be done when finalizing.
And it's clear you are trying to make the "finalizer mode" work, cause that's the new one added.

Not sure what to say, maybe making sure the dispose itself is called correctly would fix the leak much better.







---
In reply to: [193584159](https://github.com/Microsoft/react-native-windows/pull/1862#discussion_r193584159) [](ancestors = 193584159)

Reply:
Ok,
I’ll flip it back to checking the disposing book. I agree that Dispose() should be called manually first, the dtor seems like a backup.